src/storage/S3Wrapper.cpp

```diff
@@ -0,0 +1,227 @@
+            const auto& err = response.GetError();                             \
+            SPDLOG_ERROR("S3 error: {} {}",                                    \
+                         err.GetExceptionName().c_str(),                       \
+                         err.GetMessage().c_str());                            \
+            throw std::runtime_error("S3 error");                              \
+        }                                                                      \
+    }
+
+std::shared_ptr<AWSCredentialsProvider> getCredentialsProvider()
+{
+    return Aws::MakeShared<ProfileConfigFileAWSCredentialsProvider>("local");
+}
+
+ClientConfiguration getClientConf(long timeout)
+{
+    // There are a couple of conflicting pieces of info on how to configure
+    // the AWS C++ SDK for use with minio:
+    // https://stackoverflow.com/questions/47105289/how-to-override-endpoint-in-aws-sdk-cpp-to-connect-to-minio-server-at-localhost
+    // https://github.com/aws/aws-sdk-cpp/issues/587
+    ClientConfiguration config;
+
+    conf::FaasmConfig& faasmConf = conf::getFaasmConfig();
+
+    config.region = "";
+    config.verifySSL = false;
+    config.endpointOverride = faasmConf.s3Host + ":" + faasmConf.s3Port;
+    config.connectTimeoutMs = S3_CONNECT_TIMEOUT_MS;
+    config.requestTimeoutMs = timeout;
+
+    // Use HTTP, not HTTPS
+    config.scheme = Aws::Http::Scheme::HTTP;
+
+    return config;
+}
+
+void initSDK()
+{
+    SPDLOG_INFO("Initialising AWS SDK");
+    Aws::InitAPI(options);
+}
+
+void cleanUpSDK()
+{
+    Aws::ShutdownAPI(options);
+}
+
+S3Wrapper::S3Wrapper()
+  : faasmConf(conf::getFaasmConfig())
+  , clientConf(getClientConf(S3_REQUEST_TIMEOUT_MS))
+  , client(AWSCredentials(faasmConf.s3User, faasmConf.s3Password),
+           clientConf,
+           AWSAuthV4Signer::PayloadSigningPolicy::Never,
+           false)
+{}
+
+void S3Wrapper::createBucket(const std::string& bucketName)
+{
+    SPDLOG_DEBUG("Creating bucket {}", bucketName);
+    auto request = reqFactory<CreateBucketRequest>(bucketName);
+    auto response = client.CreateBucket(request);
+
+    if (!response.IsSuccess()) {
+        const auto& err = response.GetError();
+
+        auto errType = err.GetErrorType();
+        if (errType == Aws::S3::S3Errors::BUCKET_ALREADY_OWNED_BY_YOU ||
+            errType == Aws::S3::S3Errors::BUCKET_ALREADY_EXISTS) {
+            SPDLOG_DEBUG("Bucket already exists {}", bucketName);
+        } else {
+            CHECK_ERRORS(response);
+        }
+    }
+}
+
+std::vector<std::string> S3Wrapper::listBuckets()
+{
+    SPDLOG_DEBUG("Listing buckets");
+    auto response = client.ListBuckets();
+    CHECK_ERRORS(response);
+
+    Aws::Vector<Bucket> bucketObjects = response.GetResult().GetBuckets();
+
+    std::vector<std::string> bucketNames;
+    for (auto const& bucketObject : bucketObjects) {
+        const Aws::String& awsStr = bucketObject.GetName();
+        bucketNames.emplace_back(awsStr.c_str());
```

Comment:
Very minor, but given that `bucketNames` is a vector of `std::string`s and `emplace_back` constructs them in place, the following would feel more natural to me:
```cpp
bucketNames.emplace_back(awsStr.c_str(), awsStr.size());
```

Reply:
Yeah I guess it's equivalent provided the strings are properly null-terminated (albeit with a miniscule performance difference required to locate that null terminator). In this case they're coming from the `Aws::String` class which will ensure proper null termination. I'll switch in the next PR.